pkg/cmd/roachtest/tests/mixed_version_c2c.go

```diff
@@ -0,0 +1,135 @@
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/mixedversion"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+)
+
+func registerC2CMixedVersions(r registry.Registry) {
+
+	sp := replicationSpec{
+		srcNodes: 4,
+		dstNodes: 4,
+		// The timeout field ensures the c2c roachtest driver behaves properly.
+		timeout:                   10 * time.Minute,
+		workload:                  replicateKV{readPercent: 0, debugRunDuration: 1 * time.Minute, maxBlockBytes: 1, initWithSplitAndScatter: true},
+		additionalDuration:        0 * time.Minute,
+		cutover:                   30 * time.Second,
+		skipNodeDistributionCheck: true,
+		suites:                    registry.Suites(registry.Nightly),
+	}
+
+	r.Add(registry.TestSpec{
+		Name:             "c2c/mixed-versions",
+		Owner:            registry.OwnerDisasterRecovery,
+		Cluster:          r.MakeClusterSpec(sp.dstNodes + sp.srcNodes + 1),
+		CompatibleClouds: sp.clouds,
+		Suites:           registry.Suites(registry.Nightly),
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			runC2CMixedVersions(ctx, t, c, sp)
+		},
+	})
+}
+
+// TODO (msbutler): schedule upgrades during initial scan and cutover.
+func runC2CMixedVersions(ctx context.Context, t test.Test, c cluster.Cluster, sp replicationSpec) {
+
+	sourceMvt := mixedversion.NewTest(ctx, t, t.L(), c, c.Range(1, sp.srcNodes),
+		mixedversion.MinimumSupportedVersion("v23.2.0"),
+		mixedversion.AlwaysUseLatestPredecessors)
+
+	destMvt := mixedversion.NewTest(ctx, t, t.L(), c, c.Range(sp.srcNodes+1, sp.srcNodes+sp.dstNodes),
+		mixedversion.MinimumSupportedVersion("v23.2.0"),
+		mixedversion.AlwaysUseLatestPredecessors,
+	)
+
+	rd := makeReplicationDriver(t, c, sp)
+	cleanup := rd.setupC2C(ctx, t, c)
+	defer cleanup()
+
+	var pgUrl chan *url.URL
+
+	sourceTenantName := "source"
+	destTenantName := "destination"
+
+	// Start Replication
+	sourceMvt.OnStartup("start source", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		srcNode := c.Node(1)
+		srcClusterSetting := install.MakeClusterSettings()
+		addr, err := c.ExternalPGUrl(ctx, t.L(), srcNode, roachprod.PGURLOptions{})
+		if err != nil {
+			return err
+		}
+		pgURL, err := copyPGCertsAndMakeURL(ctx, t, c, srcNode, srcClusterSetting.PGUrlCertsDir, addr[0])
+		if err != nil {
+			return err
+		}
+		pgUrl <- pgURL
+		return nil
+	})
+	destMvt.OnStartup("start replication", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		sourcePGURL := <-pgUrl
+		h.System.Exec(r, fmt.Sprintf("CREATE TENANT %q FROM REPLICATION OF %q ON '%s'", destTenantName, sourceTenantName, sourcePGURL.String()))
```

Comment:
is there a way to get and set the app tenant pointing to `h.Tenant`, the application tenant that `mvt` will automagically upgrade? Ideally, the replicating tenant created for PCR and the source tenant can be manipulated by the their respective mvt drivers.

Reply:
Discussed offline, but for visibility: I don't think having tests being able to directly set `h.Tenant` at test runtime is a good idea. The test plan will have included steps to set up the original tenant (which will look confusing), not to mention that other concurrent operations could be running and swapping the tenant live would be error prone.

Possibilities to deal with this: the framework could expose a way for tests to customize tenant creation (in which case this test could include the `PCR` part of `CREATE TENANT`). Or tenant management could happen "out of band," which means the test will have to call `SET CLUSTER SETTING version` on the relevant tenant explicitly.

I'll prototype the former, but for now if we don't want to get blocked by this, we can opt for the second option and manage tenants "manually".